src/learning-catalogue/index.ts

```diff
@@ -0,0 +1,66 @@
+import {AxiosInstance} from 'axios'
+import {Course} from './model/course'
+import {CourseFactory} from './model/factory/courseFactory'
+import {PageResults} from './model/pageResults'
+import {LearningCatalogueConfig} from './learningCatalogueConfig'
+
+export class LearningCatalogue {
+	private _http: AxiosInstance
+	private _config: LearningCatalogueConfig
+	private _courseFactory: CourseFactory
+
+	constructor(http: AxiosInstance, config: LearningCatalogueConfig) {
+		this._http = http
+		this._config = config
+		this._courseFactory = new CourseFactory()
+	}
+
+	async listAll(
+		page: number = 0,
+		size: number = 10
+	): Promise<PageResults<Course>> {
+		try {
+			const response = await this._http.get(
+				`${this._config.url}/courses?page=${page}&size=${size}`,
+				{
+					auth: {
+						username: this._config.username,
+						password: this._config.password,
+					},
+				}
+			)
+
+			response.data.results = (response.data.results || []).map(
+				this._courseFactory.create
+			)
+
+			return response.data as PageResults<Course>
+		} catch (e) {
+			throw new Error(`Error listing all courses - ${e}`)
+		}
+	}
+
+	get http(): AxiosInstance {
+		return this._http
+	}
+
+	set http(value: AxiosInstance) {
+		this._http = value
+	}
+
+	get config(): LearningCatalogueConfig {
+		return this._config
+	}
+
+	set config(value: LearningCatalogueConfig) {
+		this._config = value
+	}
+
+	get courseFactory(): CourseFactory {
+		return this._courseFactory
+	}
+
+	set courseFactory(value: CourseFactory) {
+		this._courseFactory = value
+	}
```

Comment:
The public interface of your classes should define what interactions are allowed with your objects - i.e. do you want the configuration of a LearningCatalogue to be modified after the object is created? Should another entity be able to retrieve the specific CourseFactory the LearningCatalogue is using? When another entity changes the AxiosInstance the LearningCatalogue is using do we want to let them set it to null?

Don't add code (like getters and setters) for the sake of having them, or because you think you might need them in the future. And consider how they should be used - what we're saying OK to do to and with this object by how we've implemented them.

Reply:
Agreed. I've removed getters and setters for config and http  and the getter for courseFactory. The setter for course factory is needed to override with a mock in testing but id there's a better approach to this I'm happy to change. I appreciate to looks a little awkward.